Automattic/sensei
# Update blocks architecture

Close #3915

## Changes proposed in this Pull Request

* Since we need to share some components between different CPT, this PR purpose is to update the blocks architecture, standardizing namings, and folders.
* This PR also makes the contact teacher available to lessons and courses.
* Now it's very easy to add the Restricted Course to lessons too because @gkaragia already prepared the block for that and this new structure allows us to do it easily. But there are some [other reasons](https://github.com/Automattic/sensei/pull/3849#issue-539381303) that @gkaragia found to not add it to the lessons yet.

**The approach implemented in this PR is:**

In summary, we'll have assets for each CPT (or another type of grouping, if needed), and assets for _"Shared"_ blocks (blocks that can be imported in different situations).

**Pros:**
- Easy to manage.
- Don't duplicate code in different assets.
- Organized.

**Cons:**
- In the future, when we have a big lib of shared blocks, if we need only one of them, we'll import unnecessary stuff.
- Currently, our approach to register blocks for specific CPT is in the server. We need to create an approach to check it in the frontend too, so we can import shared blocks, and register only the ones we want.

**Other possible approaches:**
- Each block has its assets.
- Remove the shared assets, and import them as dependencies of each CPT (duplicating for each new case we need), based on where it'll be used.

### Folders architecture

- `asssets/blocks`
    - `[CONTEXT]-blocks`: Context folder, usually CPT, but could be other types. Eg. `course-blocks`, `lesson-blocks`, `shared-blocks`
        - `[BLOCK_NAME]-block` (Folder to each block - Ther suffix purpose is identifying block folders easily, to differentiate from other dependencies)
        - `[SUBCONTEXT]`: Group of multiple blocks, like Lesson Actions, and Course Outline
            - `[BLOCK_NAME]-block`: Blocks grouped by subcontext.

At each folder level, we can add other dependency folders.

**UPDATE: See update in the following discussions.** 

### Blocks files naming (inspired by some of our blocks and Gutenberg)

- `block.json`: Block metadata to be used in frontend and backend.
- `index.js`: Export component object.
- `edit.js`: Block `edit` function.
- `save.js`: Block `save` function.
- `settings.js`: Component with settings details, imported and instanced in the block `edit` render.
- `frontend.js`: Script imported only in the frontend, if needed.
- `style.scss`: Styles imported in editor and frontend.
- `editor.scss`: Styles imported only in the editor.
- `frontend.scss`: Styles imported only in the frontend.
- ...Other needed dependencies

**UPDATE: See update in the following discussions.** 

### _Named export_ vs _Default export_

This PR also standardizes some exports to _default export_ when it's the main purpose of the file. And it also renames some functions to match the file name.

## Testing instructions

* Just make sure the blocks continue working properly in the editor and in the frontend for courses and lessons.
* Also make sure you can add the Contact Teacher block in Lessons and Courses, and it works properly.
